### ISSUE TITLE
Readline crashes when accessing lines of history longer than three lines

### ISSUE THREAD

**Issue by wwboynton** [ConnorAtherton/rb-readline#120]
Tested on Ruby 1.9, 2.2.1

### Steps to reproduce

1. Start a simple readline loop, load history into it containing an entry over three lines long (relative to terminal columns)
2. hit the up arrow key until you arrive at the entry

#### Result

    #<TypeError: no implicit conversion from nil to integer>
    no implicit conversion from nil to integer /Users/wwboynton/.rvm/gems/ruby-1.9.3-p392/gems/rb-readline-0.5.2/lib/rbreadline.rb:2735:in `[]'
    /Users/wwboynton/.rvm/gems/ruby-1.9.3-p392/gems/rb-readline-0.5.2/lib/rbreadline.rb:2735:in `update_line'
    /Users/wwboynton/.rvm/gems/ruby-1.9.3-p392/gems/rb-readline-0.5.2/lib/rbreadline.rb:3535:in `rl_redisplay'
    /Users/wwboynton/.rvm/gems/ruby-1.9.3-p392/gems/rb-readline-0.5.2/lib/rbreadline.rb:4676:in `_rl_internal_char_cleanup'
    /Users/wwboynton/.rvm/gems/ruby-1.9.3-p392/gems/rb-readline-0.5.2/lib/rbreadline.rb:4737:in `readline_internal_charloop'
    /Users/wwboynton/.rvm/gems/ruby-1.9.3-p392/gems/rb-readline-0.5.2/lib/rbreadline.rb:4801:in `readline_internal'
    /Users/wwboynton/.rvm/gems/ruby-1.9.3-p392/gems/rb-readline-0.5.2/lib/rbreadline.rb:4823:in `readline'
    /Users/wwboynton/.rvm/gems/ruby-1.9.3-p392/gems/rb-readline-0.5.2/lib/readline.rb:45:in `readline'

###Potential Solutions

For what it's worth, I've managed to fix this by replacing the vis_pos method in the RbReadline module like so:

    def vis_pos(line)
      @vis_lbreaks[line] || 0
    end

I'll be darned if I finished diving through all the lines of code required to determine the implications of this are sound. All I know is, where once was nil, now is a value. And my program now appropriately handles history lines of any length. 

If this sounds okay, I'll fork and submit a pull request with the change. If this is an awful idea, feel free to shed some light on the 8000+ lines I'm looking at and I'd be happy to try to write a more appropriate solution.

**Comment by wwboynton** [comment]
So thanks to @ConnorAtherton, my changes got pulled, which is super spiffy, but if we could push this to rubygems that would fix the issue in the release version.

Thanks! :)

**Comment by ConnorAtherton** [comment]
Thanks @wwboynton, 0.5.3 has been released :) :+1:

**Comment by cben** [comment]
I'm still seeing this on 0.5.3, ruby 2.2.4p230, on Fedora 23.
With the long line being the last line of my .irb_history, it crashes as soon as open irb and press Up arrow:

```
bpaskinc@reserved-2-248 ~> resize
COLUMNS=70;
LINES=32;
export COLUMNS LINES;
bpaskinc@reserved-2-248 ~> nano ~/.irb_history 
bpaskinc@reserved-2-248 ~> cat ~/.irb_history
puts 'hello'
QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ
QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ
QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ
QQQQQQQQQQQQQQQQQQQ
bpaskinc@reserved-2-248 ~> irb
i/usr/local/lib/ruby/gems/2.2.0/gems/rb-readline-0.5.3/lib/rbreadline.
rb:2742:in `[]': no implicit conversion from nil to integer (TypeError
)
	from /usr/local/lib/ruby/gems/2.2.0/gems/rb-readline-0.5.3/lib
/rbreadline.rb:2742:in `update_line'
	from /usr/local/lib/ruby/gems/2.2.0/gems/rb-readline-0.5.3/lib
/rbreadline.rb:3542:in `rl_redisplay'
	from /usr/local/lib/ruby/gems/2.2.0/gems/rb-readline-0.5.3/lib
/rbreadline.rb:4683:in `_rl_internal_char_cleanup'
	from /usr/local/lib/ruby/gems/2.2.0/gems/rb-readline-0.5.3/lib
/rbreadline.rb:4744:in `readline_internal_charloop'
	from /usr/local/lib/ruby/gems/2.2.0/gems/rb-readline-0.5.3/lib
/rbreadline.rb:4808:in `readline_internal'
	from /usr/local/lib/ruby/gems/2.2.0/gems/rb-readline-0.5.3/lib
/rbreadline.rb:4830:in `readline'
	from /usr/local/lib/ruby/gems/2.2.0/gems/rb-readline-0.5.3/lib
/readline.rb:45:in `readline'
	from /usr/local/lib/ruby/2.2.0/irb/input-method.rb:150:in `get
s'
	from /usr/local/lib/ruby/2.2.0/irb.rb:469:in `block (2 levels)
 in eval_input'
	from /usr/local/lib/ruby/2.2.0/irb.rb:623:in `signal_status'
	from /usr/local/lib/ruby/2.2.0/irb.rb:468:in `block in eval_in
put'
	from /usr/local/lib/ruby/2.2.0/irb/ruby-lex.rb:188:in `call'
	from /usr/local/lib/ruby/2.2.0/irb/ruby-lex.rb:188:in `buf_inp
ut'
	from /usr/local/lib/ruby/2.2.0/irb/ruby-lex.rb:103:in `getc'
	from /usr/local/lib/ruby/2.2.0/irb/slex.rb:205:in `match_io'
	from /usr/local/lib/ruby/2.2.0/irb/slex.rb:75:in `match'
	from /usr/local/lib/ruby/2.2.0/irb/ruby-lex.rb:282:in `token'
	from /usr/local/lib/ruby/2.2.0/irb/ruby-lex.rb:264:in `lex'
	from /usr/local/lib/ruby/2.2.0/irb/ruby-lex.rb:235:in `block (
2 levels) in each_top_level_statement'
	from /usr/local/lib/ruby/2.2.0/irb/ruby-lex.rb:231:in `loop'
	from /usr/local/lib/ruby/2.2.0/irb/ruby-lex.rb:231:in `block i
n each_top_level_statement'
	from /usr/local/lib/ruby/2.2.0/irb/ruby-lex.rb:230:in `catch'
	from /usr/local/lib/ruby/2.2.0/irb/ruby-lex.rb:230:in `each_to
p_level_statement'
	from /usr/local/lib/ruby/2.2.0/irb.rb:485:in `eval_input'
	from /usr/local/lib/ruby/2.2.0/irb.rb:395:in `block in start'
	from /usr/local/lib/ruby/2.2.0/irb.rb:394:in `catch'
	from /usr/local/lib/ruby/2.2.0/irb.rb:394:in `start'
	from /usr/local/bin/irb:11:in `<main>'
```

Now if I reduce the font sufficiently, it won't crash:
```
bpaskinc@reserved-2-248 ~ [1]> resize
COLUMNS=146;
LINES=69;
export COLUMNS LINES;
bpaskinc@reserved-2-248 ~> irb
irb(main):001:0> QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ
```

**Comment by mconigliaro** [comment]
Same problem here on rb-readline 0.5.3 and Ruby 2.3.0. In my case, it gets triggered when passing a long string to [Thor](https://github.com/erikhuda/thor)'s `ask?` method. I can work around it by resizing my terminal window.

```
D/Users/Mike/.chefdk/gem/ruby/2.3.0/gems/rb-readline-0.5.3/lib/rbreadline.rb:2742:in `[]': no implicit conversion from nil to integer (TypeError)'t parse safe navigation operator, remove cancel campaign link, Fix a bug where waving/cloning offers with a rule that alrea       from /Users/Mike/.chefdk/gem/ruby/2.3.0/gems/rb-readline-0.5.3/lib/rbreadline.rb:2742:in `update_line'
	from /Users/Mike/.chefdk/gem/ruby/2.3.0/gems/rb-readline-0.5.3/lib/rbreadline.rb:3542:in `rl_redisplay'
	from /Users/Mike/.chefdk/gem/ruby/2.3.0/gems/rb-readline-0.5.3/lib/rbreadline.rb:3960:in `readline_internal_setup'
	from /Users/Mike/.chefdk/gem/ruby/2.3.0/gems/rb-readline-0.5.3/lib/rbreadline.rb:4807:in `readline_internal'
	from /Users/Mike/.chefdk/gem/ruby/2.3.0/gems/rb-readline-0.5.3/lib/rbreadline.rb:4830:in `readline'
	from /Users/Mike/.chefdk/gem/ruby/2.3.0/gems/rb-readline-0.5.3/lib/readline.rb:45:in `readline'
	from /Users/Mike/.chefdk/gem/ruby/2.3.0/gems/thor-0.19.4/lib/thor/line_editor/readline.rb:20:in `readline'
	from /Users/Mike/.chefdk/gem/ruby/2.3.0/gems/thor-0.19.4/lib/thor/line_editor.rb:7:in `readline'
	from /Users/Mike/.chefdk/gem/ruby/2.3.0/gems/thor-0.19.4/lib/thor/shell/basic.rb:410:in `ask_simply'
	from /Users/Mike/.chefdk/gem/ruby/2.3.0/gems/thor-0.19.4/lib/thor/shell/basic.rb:80:in `ask'
	from /Users/Mike/.chefdk/gem/ruby/2.3.0/gems/thor-0.19.4/lib/thor/shell.rb:59:in `ask'
```

**Comment by kplimack** [comment]
this fixed it
```
MacBook-Pro-2:knife-derp jplimack$ gem update thor rb-readline
Updating installed gems
Updating rb-readline
Fetching: rb-readline-0.5.4.gem (100%)
Successfully installed rb-readline-0.5.4
Gems updated: rb-readline
Gems already up-to-date: thor
```

**Comment by cben** [comment]
Works for me as well now on Linux (fedora 24) with rb-readline 0.5.4, but doesn't look like it was the 0.5.3 -> 0.5.4:
- commit log doesn't look like anything related was fixed recently.
- `gem uninstall rb-readline -v 0.5.4; gem install rb-readline -v 0.5.3` and it still works! (and confirmed with strace 0.5.3 is what's loaded).
- under a big bundler-using project, forcing 0.5.3 it does crash, but updating to 0.5.4, up arrow doesn't work at all, line remains empty!? (but commands are appended to .irb_history)
- cleaner experiment: inside rb-readline/ git checkout, `bundle install; bundle exec irb` => v0.5.3 works (no crash!), master up arrow doesn't show anything.

Anyway sounds like it wasn't a strictly rb-readline bug, but some interaction with other gems...

I'll report the master problem separately.

**Comment by cben** [comment]
OK, sorry for noise, here's a clearer picture.
Works for me on ruby 2.3.1 (via rbenv, on fedora 24) with both rb-readline 0.5.4 and 0.5.3.  Bisecting confirms it was indeed fixed by @wwboynton in #121, should be OK since 0.5.3 — as expected.
However with ruby 2.2.5, it still fails on 0.5.3, 0.5.4, and master.

Minimal repro in rb-readline/ directory:
```
bundle install; bundle exec ruby -e 'require "readline"; Readline::HISTORY.push("X"*400); puts Readline.readline("> ")'
# now press up arrow and see if it crashes
```

on ruby 2.2.5:
```
>/home/bpaskinc/rb-readline/lib/rbreadline.rb:2742:in `[]': no implicit conversion from nil to integer (TypeError)XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX       from /home/bpaskinc/rb-readline/lib/rbreadline.rb:2742:in `update_line'
	from /home/bpaskinc/rb-readline/lib/rbreadline.rb:3542:in `rl_redisplay'
	from /home/bpaskinc/rb-readline/lib/rbreadline.rb:4683:in `_rl_internal_char_cleanup'
	from /home/bpaskinc/rb-readline/lib/rbreadline.rb:4744:in `readline_internal_charloop'
	from /home/bpaskinc/rb-readline/lib/rbreadline.rb:4808:in `readline_internal'
	from /home/bpaskinc/rb-readline/lib/rbreadline.rb:4830:in `readline'
	from /home/bpaskinc/rb-readline/lib/readline.rb:45:in `readline'
	from -e:1:in `<main>'
```
